### PR TITLE
fix(storage): fast compactor judge whether using streaming uploader

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
close https://github.com/risingwavelabs/risingwave/issues/13726

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/hummock/compactor/fast_compactor_runner.rs [hunk 1]
```diff
@@ -279,7 +279,7 @@ pub struct CompactorRunner {
     right: Box<ConcatSstableIterator>,
     task_id: u64,
     executor: CompactTaskExecutor<
-        RemoteBuilderFactory<StreamingSstableWriterFactory, BlockedXor16FilterBuilder>,
+        RemoteBuilderFactory<UnifiedSstableWriterFactor, BlockedXor16FilterBuilder>,
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/fast_compactor_runner.rs:282]: typo: should be `xxxFactory`

File: src/storage/src/hummock/sstable_store.rs [hunk 2]
```diff
@@ -918,6 +923,59 @@ impl StreamingSstableWriterFactory {
     }
 }
 
+pub enum UnifiedSstableWriterFactor {
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable_store.rs:926]: Given that we already introduced `UnifiedSstableWriter`, we don't need to introduce a enum for `UnifiedSstableWriterFactory`. We can just have:
```rust
pub struct UnifiedSstableWriterFactory {
    sstable_store: SstableStoreRef,
}

#[async_trait::async_trait]
impl SstableWriterFactory for UnifiedSstableWriterFactory {
    type Writer = UnifiedSstableWriter;

    async fn create_sst_writer(
        &mut self,
        object_id: HummockSstableObjectId,
        options: SstableWriterOptions,
    ) -> HummockResult<Self::Writer> {
        if self.sstable_store.store().support_streaming_upload() {
            ...
            Ok(UnifiedSstableWriter::StreamingSstableWriter(
                streaming_uploader_writer,
            ))
        } else {
            ...
            Ok(UnifiedSstableWriter::BatchSstableWriter(
                batch_uploader_writer,
            ))
        }
    }
}
```

In this case, compactor doesn't need to check `support_streaming_upload` explicitly

Comment (wcy-fdu) [reply src/storage/src/hummock/sstable_store.rs]: Indeed, changed. Next pr will try to refactor [this logic](https://github.com/risingwavelabs/risingwave/blob/69604f2873b37714495b954fead0f8a9cc78a952/src/storage/src/hummock/compactor/mod.rs#L150) as well.

File: src/storage/src/hummock/compactor/fast_compactor_runner.rs [hunk 3]
```diff
@@ -313,7 +313,8 @@ impl CompactorRunner {
             split_weight_by_vnode: task.split_weight_by_vnode,
             use_block_based_filter: true,
         };
-        let factory = StreamingSstableWriterFactory::new(context.sstable_store.clone());
+        let factory = UnifiedSstableWriterFactory::new(context.sstable_store.clone());
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/fast_compactor_runner.rs:316]: We can change the [logic ](https://github.com/risingwavelabs/risingwave/blob/e6d76a2cd5d0a469ab5a50c50c998f93db827f5d/src/storage/src/hummock/compactor/mod.rs#L144) in normal compaction runner as well.

**Review by hzxa21 — APPROVED** [review]
LGTM

File: src/storage/src/hummock/compactor/fast_compactor_runner.rs (hunk 3 above)
Comment (wcy-fdu) [reply src/storage/src/hummock/compactor/fast_compactor_runner.rs:316]: Yes, will do it in next pr.

**Pull request merged**